be/src/exec/vectorized/hdfs_scanner_text.cpp

```diff
@@ -47,68 +47,64 @@ class HdfsScannerCSVReader : public CSVReader {
     _offset = offset;
     _remain_length = remain_length;
     _should_stop_scan = false;
+    _should_stop_next = false;
     _buff.skip(_buff.limit() - _buff.position());
+    return Status::OK();
 }
 
 Status HdfsScannerCSVReader::next_record(Record* record) {
-    if (_should_stop_scan) {
-        return Status::EndOfFile("Should stop for this reader!");
+    if (_should_stop_next) {
+        return Status::EndOfFile("");
+    }
+    Status st = CSVReader::next_record(record);
+    // We should still read if remain_length is zero(we stop right at row delimiter)
+    // because next scan range will skip a record till row delimiter.
+    // so it's current reader's responsibility to consume this record.
+    size_t consume = record->size + _row_delimiter_length;
+    if (_remain_length < consume) {
+        _should_stop_next = true;
+    } else {
+        _remain_length -= consume;
     }
-    return CSVReader::next_record(record);
+    return st;
 }
 
 Status HdfsScannerCSVReader::_fill_buffer() {
-    if (_should_stop_scan || _offset >= _file_length) {
-        return Status::EndOfFile("HdfsScannerCSVReader");
+    if (_should_stop_scan) {
+        return Status::EndOfFile("");
     }
 
     DCHECK(_buff.free_space() > 0);
-    Slice s;
-    if (_remain_length <= 0) {
-        s = Slice(_buff.limit(), _buff.free_space());
-    } else {
-        size_t slice_len = _remain_length;
-        s = Slice(_buff.limit(), std::min(_buff.free_space(), slice_len));
-    }
+    Slice s = Slice(_buff.limit(), _buff.free_space());
+
     // It's very critical to call `read` here, because underneath of RandomAccessFile
     // maybe its a SequenceFile because we support to read compressed text file.
     // For uncompressed text file, we can split csv file into chunks and process chunks parallelly.
     // For compressed text file, we only can parse csv file in sequential way.
     ASSIGN_OR_RETURN(s.size, _file->read(s.data, s.size));
     _offset += s.size;
-    _remain_length -= s.size;
-
     _buff.add_limit(s.size);
-    auto n = _buff.available();
+
     if (s.size == 0) {
-        if (n == 0) {
-            // Has reached the end of file and the buffer is empty.
-            _should_stop_scan = true;
-            LOG(INFO) << "Reach end of file!";
-            return Status::EndOfFile(_file->filename());
-        } else if (n < _row_delimiter_length || _buff.find(_row_delimiter, n - _row_delimiter_length) == nullptr) {
+        size_t n = _buff.available();
+        _should_stop_scan = true;
+        if (n >= _row_delimiter_length && _buff.find(_row_delimiter, n - _row_delimiter_length) == nullptr) {
             // Has reached the end of file but still no record delimiter found, which
-            // is valid, according the RFC, add the record delimiter ourself.
+            // is valid, according the RFC, add the record delimiter ourself. ONLY IF we have space.
             for (char ch : _row_delimiter) {
                 _buff.append(ch);
```

Comment:
will this exceed buf size?

Reply:
no, we have precheck

```
       size_t n = _buff.available();
        _should_stop_scan = true;
        if (n >= _row_delimiter_length && _buff.find(_row_delimiter, n - _row_delimiter_length) == nullptr) {
```

Reply:
'available' means how many content you have in the buffer. Not how many left space you have in the buffer

Reply:
Yes, you are right. We should check if there is any enough space for row delimiter.